engines/hdb/map.cpp

```diff
@@ -0,0 +1,1209 @@
+ *
+ */
+
+#include "hdb/hdb.h"
+#include "hdb/ai.h"
+#include "hdb/file-manager.h"
+#include "hdb/gfx.h"
+#include "hdb/map.h"
+
+namespace HDB {
+
+Map::Map() {
+
+	if (g_hdb->isPPC()) {
+		_screenXTiles = 9;
+		_screenYTiles = 11;
+		_screenTileWidth = 8;
+		_screenTileHeight = 10;
+	} else {
+		_screenXTiles = 17;
+		_screenYTiles = 16;
+		_screenTileWidth = 16;
+		_screenTileHeight = 16;
+	}
+
+	_mapLoaded = false;
+
+	_animCycle = 0;
+
+	_numForegrounds = _numGratings = 0;
+
+	_mapExplosions = NULL;
+	_mapExpBarrels = NULL;
+	_mapLaserBeams = NULL;
+
+	_background = NULL;
+	_foreground = NULL;
+	_iconList = NULL;
+}
+
+Map::~Map() {
+	delete[] _background;
+	delete[] _foreground;
+	delete[] _iconList;
+
+	free(_mapExplosions);
+	free(_mapExpBarrels);
+	free(_mapLaserBeams);
+}
+
+void Map::save(Common::OutSaveFile *out) {
+
+	int i;
+
+	// Save Map Dimensions and Offsets
+	out->writeSint32LE(_mapX);
+	out->writeSint32LE(_mapY);
+	out->writeSint32LE(_width);
+	out->writeSint32LE(_height);
+	out->writeSint32LE(_mapTileX);
+	out->writeSint32LE(_mapTileY);
+	out->writeSint32LE(_mapTileXOff);
+	out->writeSint32LE(_mapTileYOff);
+
+	// Save out all Level2 Gratings and AnimCycle
+
+	out->writeSint32LE(_numGratings);
+	for (i = 0; i < _numGratings; i++) {
+		out->writeUint16LE(_gratings[i].x);
+		out->writeUint16LE(_gratings[i].y);
+		out->writeUint16LE(_gratings[i].tile);
+	}
+
+	out->writeSint32LE(_animCycle);
+
+	// find out how many SLOW, MEDIUM & FAST bg tile anims there are and save them all out
+	out->writeUint32LE(_listBGAnimSlow.size());
+	for (i = 0; (uint)i < _listBGAnimSlow.size(); i++) {
+		out->writeUint32LE(_listBGAnimSlow[i]);
+	}
+
+	out->writeUint32LE(_listBGAnimMedium.size());
+	for (i = 0; (uint)i < _listBGAnimMedium.size(); i++) {
+		out->writeUint32LE(_listBGAnimMedium[i]);
+	}
+
+	out->writeUint32LE(_listBGAnimFast.size());
+	for (i = 0; (uint)i < _listBGAnimFast.size(); i++) {
+		out->writeUint32LE(_listBGAnimFast[i]);
+	}
+
+	// find out how many SLOW, MEDIUM & FAST fg tile anims there are and save them all out
+
+	out->writeUint32LE(_listFGAnimSlow.size());
+	for (i = 0; (uint)i < _listFGAnimSlow.size(); i++) {
+		out->writeUint32LE(_listFGAnimSlow[i]);
+	}
+
+	out->writeUint32LE(_listFGAnimMedium.size());
+	for (i = 0; (uint)i < _listFGAnimMedium.size(); i++) {
+		out->writeUint32LE(_listFGAnimMedium[i]);
+	}
+
+	out->writeUint32LE(_listFGAnimFast.size());
+	for (i = 0; (uint)i < _listFGAnimFast.size(); i++) {
+		out->writeUint32LE(_listFGAnimFast[i]);
+	}
+
+	// save map data
+	for (i = 0; i < _width * _height; i++)
+		out->writeSint32LE(_background[i]);
+	for (i = 0; i < _width * _height; i++)
+		out->writeSint32LE(_foreground[i]);
+	for (i = 0; i < _width * _height; i++)
+		out->writeByte(_mapExplosions[i]);
+	for (i = 0; i < _width * _height; i++)
+		out->writeByte(_mapExpBarrels[i]);
+	for (i = 0; i < _width * _height; i++)
+		out->writeByte(_mapLaserBeams[i]);
```

Comment:
These write consecutive bytes, so they could be changed to out->write(foo, _width * _height)

Reply:
I'll do that.